src/components/DataTable/README.md

```diff
@@ -0,0 +1,573 @@
+  - [`filterRows`](#filterrows)
+  - [`locale`](#locale)
+- [Render Prop Function](#render-prop-function)
+  - [Prop Getters](#prop-getters)
+  - [Actions](#actions)
+  - [State](#state)
+  - [Props](#props-1)
+- [Use-cases](#use-cases)
+  - [Sorting](#sorting)
+    - [Programmatic sorting](#programmatic-sorting)
+  - [Expansion](#expansion)
+    - [Programmatic expansion](#programmatic-expansion)
+  - [Selection](#selection)
+    - [Programmatic selection](#programmatic-selection)
+  - [Filtering](#filtering)
+- [Attribution](#attribution)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Installation
+
+This component comes with any installation of the `carbon-components-react` package on NPM. You can install this package by running the following in your terminal:
+
+```bash
+npm i carbon-components carbon-components-react carbon-icons --save
+# Or, with yarn
+yarn add carbon-components carbon-components-react carbon-icons
+```
+
+## Usage
+
+You can include `DataTable` and its components by doing the following in your project:
+
+```js
+import { DataTable } from 'carbon-components-react';
+```
+
+The default export for `DataTable` also includes properties for all the `Table*` components that you will also want to use in your application. You can access them by doing either of the following:
+
+```js
+import { DataTable } from 'carbon-components-react';
+// De-structure `DataTable` directly to get local references
+const { Table, TableHead, TableHeader, TableBody, TableCell } = DataTable;
+
+// Or, just use them in your React projects by doing
+<DataTable.Table />
+<DataTable.TableHead />
+<DataTable.TableHeader />
+// ...
+```
+
+The `DataTable` component itself follows the `render` prop pattern, meaning that in order to render something to the screen you'll have to provide a `render` function to the `DataTable` component. In practice, this looks like the following:
+
+```jsx
+import { DataTable } from 'carbon-components-react';
+const {
+  TableContainer,
+  Table,
+  TableHead,
+  TableRow,
+  TableBody,
+  TableCell,
+} = DataTable;
+
+// Inside of your component's `render` method
+function App() {
+  return (
+    <DataTable rows={initialRows}
+      headers={headers}
+      render={({ rows, headers, getHeaderProps }) => (
+        <TableContainer title="DataTable">
+          <Table>
+            <TableHead>
+              <TableRow>
+                {headers.map(header => (
+                  <TableHeader {...getHeaderProps({ header })}>
+                    {header.header}
+                  </TableHeader>
+                ))}
+              </TableRow>
+            </TableHead>
+            <TableBody>
+              {rows.map(row => (
+                <TableRow key={row.id}>
+                  {row.cells.map(cell => (
+                    <TableCell key={cell.id}>{cell.value}</TableCell>
+                  ))}
+                </TableRow>
+              ))}
+            </TableBody>
```

Comment:
It seems to me like this is a lot of structural code that could be abstracted away for an easier API.

This is great for an advanced use case, but if I just need to list some sortable data, I'd much prefer to be able to do something like this:

```
import { DataTable } from 'carbon-components-react';

const headers = [
  { key: 'name', header: 'Name' },
  { key: 'date', header: 'Date' },
];

const rows = [
  { id: 1, name: 'Foo', date: '2018-01-01' },
  { id: 2, name: 'Bar', date: '2018-02-01' },
  { id: 3, name: 'Baz', date: '2018-02-06' },
];

export default () => (
  <DataTable rows={rows} headers={headers} title="DataTable />
);
```

Any modifications would require the full markup, but I think having a "composite component" to simplify the API for default use cases would be really well received and probably cut down on your issues significantly. :)

Reply:
I think having a table that just works OOTB would be amazing 🎉 

The reason it’s being built with this approach is to allow flexibility so that it can be used in any use-case, and then we can start seeing what patterns are being used and can create dedicated components for those.

It’s also nice since the way it’s structured right now, the API you’re after can be implemented in user land 😄 

Could kind of see this methodology in how it was released:

- v1 release is just presentation, notice teams implementing ad-hoc sort/expansion/etc implementations
- v2 release consolidates basic ad-hoc implementations of state/behavior
- v3 release adds advanced behavior like inline edits
- v4 release makes it easier to use common patterns/behaviors found in the platform

We didn’t start the other way because often times teams needed direct access to inner markup in order to add handlers for the behavior we didn’t implement yet.